proxy/src/source.rs

```diff
@@ -227,16 +227,15 @@ where
                     checkpoint_id: H256::random(),
                 },
             )
-            .wait()
-            .expect("osc project registration failed");
+            .wait()?;
 
-        Project {
-            id: registry_id.into(),
+        Ok(Project {
+            id: registry_id.try_into().unwrap(),
```

Comment:
I think we want to move this outside to `let id = registry_id.try_into()?;` and then this would propagate the error :)

Reply:
That makes sense to me but apparently not the compiler:

```
error[E0277]: `?` couldn't convert the error to `substrate_subxt::error::Error`
   --> src/source.rs:220:40
    |
220 |         let id = registry_id.try_into()?;
    |                                        ^ the trait `std::convert::From<std::convert::Infallible>` is not implemented for `substrate_subxt::error::Error`
```

Reply:
I _think_ that's because you need to match the error type that is expected for this function. So in this case it's `RegistryError`, but `try_into` errors with a `String`.

So, we're getting into error handling territory that's probably worth breaking down some more. The main thing to keep in mind is that we need an error type that can encapsulate any error that can be thrown. The choices we make here depend on ergonomics (and sometimes personal preference).

If we just look at `Result` and it's instance for `Try`, allowing us to use `?`, we can get an idea of what I'm alluding to above. Let's say we have two results `registry_result: Result<Foo, RegistryError>` and `application_result: Result<Bar, ApplicationError>`.

If we were to do manually matching, in place of `?`, this would look like:
```rust
let process_registry = match registry_result {
  Ok(foo) => bar(foo)
  Err(registry_err) => return Err(registry_err)
}

let process_application = match application_result {
  Ok(bar) => baz(bar),
  Err(application_err) => Err(application_err),
}
```

The thing is that upon type-checking `RegistryError` is not the same as `ApplicationError`. So we will get a type error saying so.

**BUT**, there's something else that's sneaky about `?` that I only noticed by accident. Say if you had an impl for `Into` from `RegistryError` to `ApplicationError`, `?` will use that an convert it under the hood. So if your top-level signature is `Result<Baz, ApplicationError>` it will know that `RegistryError` can be converted and do that for you. I'm not sure about the internals here.

So there are two tactics that come to mind here, but maybe people can chime in with more options.

1. Make an `enum` of the types of errors that you can run into, and if there's outside cases then you can add that to the `enum` and write an `Into` instance. You can see `radicle-surf` using this [here](https://github.com/radicle-dev/radicle-surf/blob/master/src/vcs/git.rs#L49-L56).

2. Use `String. It's easy to turn things into `String` so you can always just use that.

You may be asking why would you use one over the other? 

- `enum` captures the error cases that your functions can call (but technically can still do things like `panic`), but are hard to extend and you get into a kind of `enum` hell, matching on layer after layer.

- `String` has little information as to what the error was, you can't match on particular cases, but extends easily because of being able to just represent anything as a `String`.

Welcome to the world of explicit error handling :hugs:

Reply:
Coming from JavaScript where error handling is all about opaque strings, I'm inclined to go with the enum approach... or a ProxyError type with From methods for everything we expect?

Reply:
I think the `enum` approach is the `ProxyError` type with `From` method for `RegistryError` :)
So you'd have an `enum ProxyError` that has the specific errors that occur in Proxy, and then one of the branches is `Registry(RegistryError)`